fuse_core/include/fuse_core/autodiff_local_parameterization.h

```diff
@@ -0,0 +1,222 @@
+   * @param[out] jacobian The Jacobian in row-major order, of size GlobalSize() x LocalSize()
+   * @return True is successful, false otherwise
+   */
+  bool ComputeJacobian(
+    const double* x,
+    double* jacobian) const override;
+
+  /**
+   * @brief Generalization of the subtraction operation, implemented by the provided MinusFunctor
+   *
+   * @param[in]  x1    The value of the first variable, of size \p GlobalSize()
+   * @param[in]  x2    The value of the second variable, of size \p GlobalSize()
+   * @param[out] delta The difference between the second variable and the first, of size \p LocalSize()
+   * @return True if successful, false otherwise
+   */
+  bool Minus(
+    const double* x1,
+    const double* x2,
+    double* delta) const override;
+
+  /**
+   * @brief The Jacobian of Minus(x1, x2) w.r.t x2 evaluated at x1 = x2 = x, computed using automatic differentiation
+   * @param[in]  x        The value used to evaluate the Jacobian, of size \p GlobalSize()
+   * @param[out] jacobian The Jacobian in row-major order, of size \p LocalSize() x \p GlobalSize()
+   * @return True is successful, false otherwise
+   */
+  bool ComputeMinusJacobian(
+    const double* x,
+    double* jacobian) const override;
+
+  /**
+   * @brief The size of the variable parameterization in the nonlinear manifold
+   */
+  int GlobalSize() const override { return kGlobalSize; }
+
+  /**
+   * @brief The size of a delta vector in the linear tangent space to the nonlinear manifold
+   */
+  int LocalSize() const override { return kLocalSize; }
+
+private:
+  std::unique_ptr<PlusFunctor> plus_functor_;
+  std::unique_ptr<MinusFunctor> minus_functor_;
+};
+
+template <typename PlusFunctor, typename MinusFunctor, int kGlobalSize, int kLocalSize>
+AutoDiffLocalParameterization<PlusFunctor, MinusFunctor, kGlobalSize, kLocalSize>::AutoDiffLocalParameterization() :
+  plus_functor_(new PlusFunctor()),
+  minus_functor_(new MinusFunctor())
+{
+}
+
+template <typename PlusFunctor, typename MinusFunctor, int kGlobalSize, int kLocalSize>
+AutoDiffLocalParameterization<PlusFunctor, MinusFunctor, kGlobalSize, kLocalSize>::AutoDiffLocalParameterization(
+  PlusFunctor* plus_functor,
+  MinusFunctor* minus_functor) :
+    plus_functor_(new PlusFunctor()),
+    minus_functor_(new MinusFunctor())
+{
+}
+
+template <typename PlusFunctor, typename MinusFunctor, int kGlobalSize, int kLocalSize>
+bool AutoDiffLocalParameterization<PlusFunctor, MinusFunctor, kGlobalSize, kLocalSize>::Plus(
+  const double* x,
+  const double* delta,
+  double* x_plus_delta) const
+{
+  return (*plus_functor_)(x, delta, x_plus_delta);
+}
+
+template <typename PlusFunctor, typename MinusFunctor, int kGlobalSize, int kLocalSize>
+bool AutoDiffLocalParameterization<PlusFunctor, MinusFunctor, kGlobalSize, kLocalSize>::ComputeJacobian(
+  const double* x,
+  double* jacobian) const
+{
+  double zero_delta[kLocalSize];
+  for (int i = 0; i < kLocalSize; ++i)
```

Comment:
Doubt it's any faster, but can you use `std::fill` here?

```
std::fill(zero_delta, zero_delta + kLocalSize, 0.0);
```

Reply:
These are actually zero-initialized by the compiler. Removing the initialization entirely.
https://en.cppreference.com/w/cpp/language/zero_initialization
> double f[3]; // zero-initialized to three 0.0's

Reply:
Oh, sweet.

Reply:
Actually, maybe I'm lying. That's only for static/global-scope.

Reply:
But empty brace-initialization does work.
`double zero_delta[kLocalSize] = {};`

Reply:
Yay C++14.